File: Python/HackerRank/climbingLeaderboard.py

def climbingLeaderboard(ranked, player):
    unique_ranked = sorted(list(set(ranked)), reverse=True)
    
    results = []
    i = len(unique_ranked) - 1
    
    for score in player:
        while i >= 0 and score >= unique_ranked[i]:
            i -= 1
        
        if i < 0:
            results.append(1)
        else:
            results.append(i + 2)
            
    return results

def main():
    n = int(input())
    ranked = list(map(int, input().rstrip().split()))
    m = int(input())
    player = list(map(int, input().rstrip().split()))

    if n != len(ranked) or m != len(player):
        raise ValueError

    print(*climbingLeaderboard(ranked, player), sep='\n')

if __name__ == "__main__":
    main()

File: golang/hackerrank/arrays/climbingLeaderboard.go
package arrays
import (
	"sort"
)








// ClimbingLeaderboard calculates the player's rank after each game using Dense Ranking.
func ClimbingLeaderboard(ranked []int, player []int) []int {
	// Get unique ranked scores and sort them descending
	uniqueMap := make(map[int]bool)
	for _, r := range ranked {
		uniqueMap[r] = true
	}
	var uniqueRanked []int
	for k := range uniqueMap {
		uniqueRanked = append(uniqueRanked, k)
	}
	sort.Slice(uniqueRanked, func(i, j int) bool {
		return uniqueRanked[i] > uniqueRanked[j]
	})

	results := make([]int, 0, len(player))
	i := len(uniqueRanked) - 1

	for _, score := range player {
		for i >= 0 && score >= uniqueRanked[i] {
			i--
		}

		if i < 0 {
			results = append(results, 1)
		} else {
			results = append(results, i+2)
		}
	}

	return results
}
